### PR TITLE
Feat/allow multiple countdowns

### PR DESCRIPTION
**Pull request by wmui51**
# NOT READY FOR REVIEW
- (Edit the above to reflect status)

# Summary
- TL;DR - what's this PR for?
The Countdown component did not allow for multiple instances on a page because of the way the elements within the components had their class names set up. This updates it so that every instance should get a random class name. There is still a possibility that two instances could generate the same random class name, but that is low and if it becomes a problem we can make this configurable via Storyblok.
image

Also cleaned up some redundant code.

# Review By (Date)
- When does this need to be reviewed by?
- I dont' believe there is a rush as there isn't an active one in use.

# Criticality
- How critical is this PR on a 1-10 scale? Also see [Severity Assessment](https://stanfordits.atlassian.net/browse/D8CORE-1705).
- Not critical

# Review Tasks

## Setup tasks and/or behavior to test

1. Within a test page on Storyblok, see if you can add more than one Countdown without things getting "weird."

## Front End Validation
- [ ] Is the markup using the appropriate semantic tags and passes HTML validation?
- [ ] Cross-browser testing has been performed?
- [ ] Automated accessibility scans performed?
- [ ] Manual accessibility tests performed?
- [ ] Design is approved by @ user?

### Code
- [ ] Are the naming conventions following our standards?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided? eg (unit, behat, or codeception)

### Code security
- [ ] Are all [forms properly sanitized](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

## General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules?

# Associated Issues and/or People
- JIRA ticket(s)
- Other PRs
- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- Anyone who should be notified? (`@mention` them here)

# Resources
- [AMP Tool](https://stanford.levelaccess.net/index.php)
- [Accessibility Manual Test Script](https://docs.google.com/document/d/1ZXJ9RIUNXsS674ow9j3qJ2g1OAkCjmqMXl0Gs8XHEPQ/edit?usp=sharing)
- [HTML Validator](https://validator.w3.org/)
- [Browserstack](https://live.browserstack.com/dashboard) and link to [Browserstack Credentials](https://asconfluence.stanford.edu/confluence/display/SWS/External+Account+Credentials)

### CODE REVIEW

File: src/components/giving-tuesday/Countdown.js [hunk 1]
```diff
@@ -1,10 +1,13 @@
-import React, { useEffect, useState } from 'react';
+import React, { useEffect, useMemo, useState } from 'react';
 import classNames from 'classnames';
 import SbEditable from 'storyblok-react';
 
 import CountdownPie from './CountdownPie';
 import UseCountdown from '../../hooks/useCountdown';
 
+const generateClassName = (key) =>
+  `${key}-${Math.floor(Math.random() * 100000, 5)}`;
```
Comment (sherakama) [inline src/components/giving-tuesday/Countdown.js:9]: If you are worried about uniqueness you could use the built-in hook https://react.dev/reference/react/useId

File: src/components/giving-tuesday/Countdown.js [hunk 2]
```diff
@@ -50,71 +58,44 @@ const Countdown = ({ blok }) => {
...
+        role="timer"
+      >
+        {days > 0 && hasDays && (
           <CountdownPie
-            className="pie-seconds"
-            descriptor="seconds"
-            percent={0}
+            className={daysClassName}
+            descriptor={getDescriptorString('day', days)}
+            percent={noTime ? 0 : (days / 29) * 100}
```
Comment (sherakama) [inline src/components/giving-tuesday/Countdown.js:72]: 29 days?

**Review by sherakama — CHANGES_REQUESTED** [review]
This looks fine, just checking in on the math for the days countdown. Are you sure you want to assume 29?

File: src/components/giving-tuesday/Countdown.js (hunk 1 above)
Comment (wmui51) [reply src/components/giving-tuesday/Countdown.js:9]: i will switch it up to `useId`

File: src/components/giving-tuesday/Countdown.js (hunk 2 above)
Comment (wmui51) [reply src/components/giving-tuesday/Countdown.js:72]: good catch @sherakama, i believe that was hardcoded for the specific countdown in place. i will rework that.

File: src/components/giving-tuesday/Countdown.js (hunk 1 above)
Comment (wmui51) [reply src/components/giving-tuesday/Countdown.js:9]: actually it looks like `useId` was introduced in React18, and the site is still on 17.

**Comment by wmui51** [comment]
@sherakama actually i believe this is good to go for now. i have another incoming pr to make the time configurable via storyblok, so will add the days math updates in that pr. i have added a comment to use `useId` when it becomes available.

File: src/components/giving-tuesday/Countdown.js (hunk 1 above)
Comment (sherakama) [reply src/components/giving-tuesday/Countdown.js:10]: Oh huh, I thought we moved all our stuff to 18 already. No biggie. 

Just as a FYI, at some point, we will be replacing Gatsby for Next.js in each of our sites including this one. Just thought I would give you the heads up but nothing has started for this project yet and it will likely be some time before it does.

**Comment by sherakama** [comment]
@wmui51 Ok. Do you want to wait until the new PR and math are available or do you need this work on dev?

**Comment by wmui51** [comment]
@sherakama i don't need this on dev now, but wanted to do the two things in two separate prs to make them easier for review. also there will be a few more things i have to edit to make it fully Storyblok configurable so figured it would be best to keep things separate.

but if you think i should have it all in this pr, i can do that too.

File: src/components/giving-tuesday/Countdown.js (hunk 1 above)
Comment (wmui51) [reply src/components/giving-tuesday/Countdown.js:10]: sounds good, in my brief experience working on Next it felt similar to Gatsby. what's the reason for the switch, is Gatsby deprecated?

Comment (sherakama) [reply src/components/giving-tuesday/Countdown.js:10]: The TLDR; is that Netlify bought Gatsby (the company) and isn't making the same investment into the JS Framework. 

Gatsby is on the decline
https://2022.stateofjs.com/en-US/libraries/rendering-frameworks/

and we are using Next in other places and like it.

**Comment by sherakama** [comment]
> @sherakama i don't need this on dev now, but wanted to do the two things in two separate prs to make them easier for review. also there will be a few more things i have to edit to make it fully Storyblok configurable so figured it would be best to keep things separate.
> 
> but if you think i should have it all in this pr, i can do that too.

Separated PRs is great. I do like smaller chunks to review and I appreciate the effort. I was fishing for what to do with this code and whether to hold on the merge or to send it through. I think I'll hold on merging this and you can [PR stack](https://blog.logrocket.com/using-stacked-pull-requests-in-github/) and send in your next PR into this working branch. 

Thanks @wmui51

**Comment by wmui51** [comment]
sounds good, i will branch off this one for the additional work.

**Comment by sherakama** [comment]
@wmui51 and @forrestglick 

I have added this preview build to the dev storyblok space so you can review this functionality. Select PR #425 when choosing a preview window in Storyblok or follow this link: https://app.storyblok.com/#/me/spaces/122401/stories/0/0/397282734?env=PR+425

**Comment by sherakama** [comment]
@wmui51 and @forrestglick 

This PR has been sitting for a bit. Do you have an update for me on the status of this work?

**Comment by forrestglick** [comment]
@sherakama Thanks for the ping on this PR. I'm still a bit confused about why I can't see the date/time picker with my current timezone calculated (I see GMT-7), so it makes it more difficult to set the end time for the counter. Thus far, all our timers have been based on Pacific Time btw. If there's no way to address this, we can go with the code as-is.
The other concern is that we rolled out a page to production using this current countdown timer, so I need to understand whether there could be any impact on the existing implementation. The live prod page is: https://giving.stanford.edu/home-of-champions-challenge-2024/

**Comment by wmui51** [comment]
@forrestglick I think the current work could impact the timer that's live. I am curious about the GMT-7 in the datepicker as mine shows GMT-0 when I try on two different devices. Do you want me to scrap the Storyblok datepicker and just use an input that a user can enter the date? ie. "08-05-2024 15:59"

**Comment by forrestglick** [comment]
@wmui51 If you feel the Storyblock datepicker is the source of the problem with the timezone, I'd be ok with using a simple input to set the date/time. All our timers will be based off Pacific Time.

**Comment by yvonnetangsu** [comment]
> @forrestglick I think the current work could impact the timer that's live. I am curious about the GMT-7 in the datepicker as mine shows GMT-0 when I try on two different devices. Do you want me to scrap the Storyblok datepicker and just use an input that a user can enter the date? ie. "08-05-2024 15:59"

@wmui51 @forrestglick  I wonder if that could be related to your personal account settings in Storyblok? There is a time zone preference field:
Screenshot 2024-07-29 at 10 50 25 AM

**Comment by forrestglick** [comment]
Hi @yvonnetangsu – thanks for the suggestion of checking my SB account's timezone setting. I can confirm it's set to Pacific time. Not sure why I'm getting GMT-7, but if it's working correctly for you and Wilbert, I think I might just be an anomaly! I'm ok with scrapping the date/time picker. It's always felt a bit awkward to use. And knowing we have a live instance of the countdown timer (ending with the Olympics), I'd like to hold off on the prod push until that campaign is completed. Apologies for the delay in getting this wrapped up.

**Comment by wmui51** [comment]
@yvonnetangsu @sherakama I have talked to @forrestglick and we should be good to merge and deploy this for now. There aren't any live countdowns.

**Comment by sherakama** [comment]
@wmui51 if you aren't planning on using it right away but want to commit to this code before it gets stale, I can merge it into dev to let it _bake_ for a bit. It would be released on the next release of something or at the time you say so.

That work for you?

**Comment by wmui51** [comment]
@sherakama i think that makes sense. i think it being in dev will allow for further testing of the code too.

**Comment by sherakama** [comment]
Great. Thanks @wmui51

**Review by sherakama — APPROVED** [review]
GTG

**Comment by yvonnetangsu** [comment]
Thank you both!
